src/protocol/sort/apply.rs

```diff
@@ -11,77 +12,121 @@ use permutation::Permutation;
 
 #[embed_doc_image("apply", "images/sort/apply.png")]
 #[embed_doc_image("apply_inv", "images/sort/apply_inv.png")]
+#[embed_doc_image("apply", "images/sort/apply.png")]
+#[embed_doc_image("apply_inv", "images/sort/apply_inv.png")]
+
 /// Permutation reorders (1, 2, . . . , m) into (σ(1), σ(2), . . . , σ(m)).
 /// For example, if σ(1) = 2, σ(2) = 3, σ(3) = 1, and σ(4) = 0, an input (A, B, C, D) is reordered into (C, D, B, A) by σ.
 /// ![Apply steps][apply]
-pub fn apply<T: Copy + Default, S>(mut permutation: Permutation, values: &mut S)
-where
-    S: AsMut<[T]>,
-{
-    permutation.apply_slice_in_place(values);
+pub fn apply<T: Copy + Default>(permutation: &[usize], values: &mut [T]) {
+    let mut permuted = bitvec![0; permutation.len()];
+    let mut tmp: T = T::default();
+
+    for i in 0..permutation.len() {
+        if permuted[i] == false {
+            mem::swap(&mut tmp, &mut values[i]);
+            let mut pos_i = i;
+            let mut pos_j = permutation[pos_i];
+            while pos_j != i {
+                values[pos_i] = values[pos_j];
+                pos_i = pos_j;
+                pos_j = permutation[pos_i];
+                permuted.set(pos_i, true);
+            }
+            mem::swap(&mut values[pos_i], &mut tmp);
+            permuted.set(i, true);
```

Comment:
I am almost certain you don't need it as i th element is in place already and you should never visit it again.

Reply:
This is true. As an optimization I'll remove this. It just felt good to leave the `permuted` vector all ones at the end. This will leave it with a few zeroes in there.. which sort of feels yucky.. but there's no benefit to doing this. extra work so why keep it?